### PR TITLE
Feat/fix interrupts

### PR DESCRIPTION
**Pull request by thisisarjun**
### Overview

This PR introduces a significant refactor to the node architecture handling meeting detail collection within MeetingMuse, with a focus on modularity, improved prompt handling, testability, and developer experience.

image

image

#### Key Features & Changes

---

### 1. **Refactored Node Flow for Meeting Details**

- **Introduces `PromptMissingMeetingDetailsNode`:**
  - A new node class that specifically prompts the user for missing required meeting details.
  - Logic for generating and presenting the prompt is separated from the logic that gathers the user's input.
  - This node is now the explicit transition point when meeting information is incomplete.

- **Refactors `HumanScheduleMeetingMoreInfoNode`:**
  - Now only responsible for handling and processing human input after the prompt is shown.
  - No longer generates prompts itself; instead, it expects the relevant prompt to be set in the state by the previous node.
  - Improved handling of user responses, including empty or whitespace-only inputs.

- **Graph Flow:**
  - The main graph (`GraphBuilder`) is updated to insert the new prompt node before handing over to the human input node.
  - Ensures the conversation flow is:  
    `COLLECTING_INFO` → `PROMPT_MISSING_MEETING_DETAILS` → `HUMAN_SCHEDULE_MEETING_MORE_INFO` (→ back to `COLLECTING_INFO` or `END` as appropriate).

---

### 2. **Makefile: Developer Experience Improvements**

- **Colorized and Sectioned Help Output:**
  - The `help` target now outputs a colorized, categorized list of commands, with quick examples and better readability.
- **New Targets:**
  - `build-graph`: Generates a visual graph of the conversation flow (see below).
  - `debug-node`, `debug-chatbot`: For isolated node and chatbot interaction debugging.
  - `qa`, `dev-setup`, `ci-test`: Composite quality/dev commands.
- **Standardized Target Names:**
  - Targets like `shell`, `update` now have simpler names.
  - Old aliases are removed or redirected for consistency.

try 
```
make help
```

---

### 3. **Visualization and Debugging**

- **Graph Visualization:**
  - Adds `tests/scripts/generate_graph.py` and a `build-graph` Makefile target to auto-generate a visual representation (`graph.png`) of the current conversation/node flow.
  - Useful for onboarding and debugging complex flows.

- **Debug Scripts:**
  - Adds and documents scripts for debugging individual nodes (`run_node_with_graph.py`) and the chatbot session (`run_chatbot.py`), with improved CLI argument handling and interruption support.

---

### 4. **Testing**

- **Comprehensive Unit Tests:**
  - Adds isolated tests for both `PromptMissingMeetingDetailsNode` and `HumanScheduleMeetingMoreInfoNode` (with various input scenarios, including edge cases).
  - Existing tests for `CollectingInfoNode` are updated to reflect the new transition logic and node names.
  - Improved mocking and dependency injection for easier test maintenance.

- **Test Code Quality:**
  - Ensures all new logic is covered, with explicit tests for error handling, fallback paths, and prompt generation.

---

### 5. **State and Model Updates**

- **State Model Refactored:**
  - MeetingMuseBotState now uses `ai_prompt_input` for storing dynamically generated prompt content for the next human input, instead of ambiguous flags.

- **Node Name Enum:**
  - Adds `PROMPT_MISSING_MEETING_DETAILS` to `NodeName`.
  - Improves docstrings and organization of node types (human vs prompt nodes).

---

### 6. **Code Quality & Maintenance**

- **Removes Dead Code:**
  - Removes the unused `meeting_review_node.py` and related references.
- **General Cleanup:**
  - Improves logging for node transitions and user input stages.
  - Modernizes Makefile with better help documentation, color, and grouped commands.

---

## How to Test

1. **Run Unit Tests:**
   ```bash
   make test
   ```

2. **Try the CLI Debug Scripts:**
   - Node isolation:
     ```bash
     make debug-node NODE_NAME=COLLECTING_INFO MESSAGE="Schedule a meeting"
     ```
   - Full chatbot:
     ```bash
     make debug-chatbot
     ```

3. **Generate and View the Conversation Graph:**
   ```bash
   make build-graph
   open graph.png
   ```

4. **Check Makefile Help:**
   ```bash
   make help
   ```

---

## Checklist

- [x] All tests passing
- [x] All new code documented and type-hinted
- [x] Makefile help updated
- [x] Graph visualization script included

---

**Reviewers:**  
Please focus on the node transition logic, test coverage, and Makefile usability. Feedback on developer experience and the clarity of graph visualizations is especially welcome.

---

### CODE REVIEW

File: src/meetingmuse/graph.py [hunk 1]
```diff
@@ -52,17 +59,19 @@ def build(self) -> StateGraph:
...
-                NodeName.COLLECTING_INFO: NodeName.COLLECTING_INFO,
+                # TODO: should go to schedule meeting node
                 NodeName.END: END,
+                NodeName.PROMPT_MISSING_MEETING_DETAILS: NodeName.PROMPT_MISSING_MEETING_DETAILS,
             }
-        )
+        )        
+        graph_builder.add_edge(self.prompt_missing_meeting_details_node.node_name, self.human_schedule_meeting_more_info_node.node_name)
+        graph_builder.add_edge(self.human_schedule_meeting_more_info_node.node_name, self.collecting_info_node.node_name)
         # Add edges to END for completion
```
Comment (ismailmmd) [inline src/meetingmuse/graph.py:69]: In context with with comment - https://github.com/thisisarjun/meetingmuse/pull/5#discussion_r2233082719
Which routing technique are we taking? I see router defined here and alos `Commands` defined within Nodes

File: tests/scripts/run_chatbot.py [hunk 2]
```diff
@@ -21,30 +25,54 @@
...
+            user_intent=None,
+            meeting_details=MeetingFindings()
+        )
+    def get_last_message(self, events: dict):
+      # Skip interrupt events
+        if "__interrupt__" in events:
+            return None
+
+        # Get the state from the node result
+        for node_name, state in events.items():
```
Comment (ismailmmd) [inline tests/scripts/run_chatbot.py:48]: `node_name` is not used
```suggestion
        for state in events.values():
```

File: tests/scripts/run_chatbot.py [hunk 3]
```diff
@@ -21,30 +25,54 @@
...
+                message = state.messages[-1]
+                if message and message.type == "ai" and message.content:
+                    return message.content
+        return None
+
+
     def process_input(self, user_input: str):
         # Always add the user message and process
-        input_data = {"messages": [HumanMessage(content=user_input)]}
+        self.initial_state.messages.append(HumanMessage(content=user_input))
```
Comment (ismailmmd) [inline tests/scripts/run_chatbot.py:58]: We should add a validation here that a valid key exist for LLM API, before start processing. Wdyt?

File: tests/scripts/run_node_with_graph.py [hunk 4]
```diff
@@ -103,51 +104,57 @@ def test_single_node(node_name: NodeName, user_message: str):
...
+            user_input = input(f"{interrupt_info.value} ")
+            for resume_chunk in graph.stream(Command(resume=user_input), config):
+                logger.info(f"🆔 resume_chunk: {resume_chunk}")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Debug individual nodes in isolation")
+    parser.add_argument("--node", type=str, required=True, 
+                       choices=["CLASSIFY_INTENT", "GREETING", "COLLECTING_INFO", "CLARIFY_REQUEST", "HUMAN_SCHEDULE_MEETING_MORE_INFO", "PROMPT_MISSING_MEETING_DETAILS"],
+                       help="Node name to test")
```
Comment (ismailmmd) [inline tests/scripts/run_node_with_graph.py:143]:
```suggestion
                       choices=[node.name for node in NodeName],

```

File: src/meetingmuse/nodes/human_schedule_meeting_more_info_node.py [hunk 5]
```diff
@@ -48,8 +29,9 @@ def node_action(self, state: MeetingMuseBotState) -> MeetingMuseBotState:
         
         # Parse human input and update meeting details
         state.messages.append(HumanMessage(content=human_input))
+        state.ai_prompt_input = None
         self.logger.info("Human input processed, continuing to collecting_info node")
-        return Command(goto="collecting_info")
+        return state
```
[the remark this answers is not in the document]
Comment (thisisarjun) [reply src/meetingmuse/nodes/human_schedule_meeting_more_info_node.py:34]: i have reverted this, so

yes, lets have two ways of routing
- routing via Command
- routing via conditional edge

File: tests/scripts/run_chatbot.py (hunk 2 above)
Comment (thisisarjun) [reply tests/scripts/run_chatbot.py:48]: nope, this wont work, events.items() gives a [tuple](https://docs.python.org/3/library/stdtypes.html#tuples), and when we do  node_name, state -> we are unpacking the tuple to node_name and state, 

for reference
```python
abc = {
    "arjun": "one",
    "da": "2"
}

for v in abc.items():
    print(abc)

for k, v in abc.items():
    print(f"{k}: {v}")
```
gives
```python
{'arjun': 'one', 'da': '2'}
{'arjun': 'one', 'da': '2'}
arjun: one
da: 2
```

File: tests/scripts/run_chatbot.py (hunk 3 above)
Comment (thisisarjun) [reply tests/scripts/run_chatbot.py:58]: since this is a debug chatbot, imo could be a little overkill?, once we have implemented the actual chatbot, we can add the validations. wdyt?

File: tests/meetingmuse/services/test_meeting_details_service.py [hunk 6]
```diff
@@ -1,6 +1,9 @@
 import pytest
-from unittest.mock import Mock, MagicMock
+from unittest.mock import Mock, MagicMock, patch
```
Comment (ismailmmd) [inline tests/meetingmuse/services/test_meeting_details_service.py:2]:
```suggestion
from unittest.mock import Mock, patch
```

File: tests/meetingmuse/services/test_meeting_details_service.py [hunk 7]
```diff
@@ -1,6 +1,9 @@
...
+
 from meetingmuse.services.meeting_details_service import MeetingDetailsService
+from meetingmuse.llm_models.hugging_face import HuggingFaceModel
+from meetingmuse.utils.logger import Logger
```
Comment (ismailmmd) [inline tests/meetingmuse/services/test_meeting_details_service.py:6]:
```suggestion
```

**Review by ismailmmd — APPROVED** [review]
